substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -243,6 +245,7 @@ pub trait TransactionExtension<Call: Dispatchable>:
 		len: usize,
 		self_implicit: Self::Implicit,
```

Comment:
outside of this PR: I think we'll also need:

```
/// Versions for which this Transaction Extension is applicable.
const VERSIONS: &'static [u8]
```

I could handle this in my metadata update PR

Reply:
I don't think we'll do it that way but rather a list of versions supported in the current `TxExtension` definition, which would end up being an `enum` like structure for each supported version. But in this format, it would not be on the `TransactionExtension` trait definition, but on the `ExtrinsicMetadata` definition, as in one for the big, tuple extension of the runtime.

Reply:
@lexnv what you mean by this?

Reply:
I've created a new PR for frame-metadata to add a few minor [changes](https://github.com/paritytech/frame-metadata/pull/86/files#diff-f37e220a086df34a79724e6c39847334c03fe634544eef13c2308eeea6dad8f2R187). The `ExtrinsicMetadata` looks like: 

```rust
pub struct ExtrinsicMetadata<T: Form = MetaForm> {
	/// Extrinsic versions.
	pub versions: Vec<u8>,
	/// The type of the address that signs the extrinsic
	pub address_ty: T::Type,
	/// The type of the extrinsic's signature.
	pub signature_ty: T::Type,
	/// A mapping of supported transaction extrinsic versions to their respective transaction extension indexes.
	///
	/// For each supported version number, list the indexes, in order, of the extensions used.
	pub transaction_extensions_by_version: BTreeMap<u8, Vec<u32>>,
	/// The transaction extensions in the order they appear in the extrinsic.
	pub transaction_extensions: Vec<TransactionExtensionMetadata<T>>,
}
```

- (1) `ExtrinsicMetadata::versions` should be [4 (legacy), 5 (new format)]. This can be done by changing the [ExtrinsicMetadata trait](https://github.com/paritytech/polkadot-sdk/pull/5732/files#diff-7be1730dfe08652baee2cb237c2d33df42c5eb85408639864c6a4a9223f0ca32R393)

- (2) `TransactionExtensionMetadata` are grouped by their extension version (currently only version 0u8 is supported). But in the future we might have multiple transaction extension versions which have different tx extensions and possibly different order of encoding

I wanted to get the information from (2) available for collection in metadata 🤔 Just to double check is my understanding of (2) correct? 🙏

Reply:
I think (1) is correct to me.

in (2) you mean that a transaction extension like `CheckWeight` will be valid for a predefined set of version? This is not the correct level. The runtime is defining multiple version of transaction extension set.

So a runtime can have:
* Transaction extensions version 33: `CheckWeight` + `Payment`.
* Transaction extensions version 34: `CheckWeight` + `Payment` + `CheckMetadataVersion`.
* Transaction extensions version 35: `Payment`.

Then an extrinsic can choose one transaction extensions version.

I believe the syntax will be something a bit like: (syntax can be different)
```rust
type UncheckedExtrinsic = UncheckedExtrinsic<Address, Signature, ((const 33, TxExtensionV33), (const 34, TxExtensionV34), (const 35, TxExtensionV35))>

type TxExtensionV33 = (CheckWeight, Payment);

//etc...
```

Reply:
Yeah 1 and 2 make sense. However, I still don't get the initial comment. Maybe what Guillaume has written above makes it more clear. The version is in the runtime and not in the extension. If we need to change extensions, we need to give them a new name.